yt/utilities/lib/cyoctree.pyx

```diff
@@ -0,0 +1,412 @@
+        self.num_particles = input_pos.shape[0]
+
+    cdef setup_ctree(self, double[:, ::1] &input_pos, int n_ref):
+        self.c_tree.n_ref = n_ref
+        self.c_tree.max_depth = 0
+
+        self.idx = np.arange(0, input_pos.shape[0], dtype=np.int32)
+        self.c_tree.idx = &self._idx[0]
+
+        reserve(&self.c_tree.nodes, 10000000)
+
+    cdef setup_bounds(self, double[:, ::1] &input_pos, left_edge=None,
+                     right_edge=None):
+        if left_edge is not None:
+            for i in range(3):
+                self.c_tree.left_edge[i] = left_edge[i]
+        else:
+            for i in range(3):
+                self.c_tree.left_edge[i] = np.amin(input_pos[:,i])
+
+        if right_edge is not None:
+            for i in range(3):
+                self.c_tree.right_edge[i] = right_edge[i]
+        else:
+            for i in range(3):
+                self.c_tree.right_edge[i] = np.amax(input_pos[:,i])
+
+    cdef setup_root(self, double[:, ::1] &input_pos):
+        cdef Node root
+        root.left_edge = self.c_tree.left_edge
+        root.right_edge = self.c_tree.right_edge
+        root.parent = -1
+        root.start = 0
+        root.end = input_pos.shape[0]*3
+        root.children = -1
+        root.node_id = 0
+        root.leaf = 1
+        root.depth = 0
+        root.leaf_id = 0
+        root.node_id = 0
+
+        # store the root in an array and make a convenient pointer
+        self.c_tree.nodes.push_back(root)
+        self.c_tree.root = &(self.c_tree.nodes[0])
+
+    cdef reset(self):
+        # reset the c tree
+        self.c_tree.nodes.clear()
+        for i in range(3):
+            self.c_tree.left_edge[i] = 0.0
+            self.c_tree.right_edge[i] = 0.0
+        self.c_tree.n_ref = 0
+
+        # reset python properties
+        self.n_ref = 0
+        self.num_octs = 0
+        self.idx = np.array([0], dtype=np.int32)
+        self.num_particles = 0
+
+    @property
+    def size_bytes(self):
+        return sizeof(Node) * self.c_tree.nodes.size()
+
+    @property
+    def max_depth(self):
+        return self.c_tree.max_depth
+
+    @property
+    def idx(self):
+        return np.asarray(self._idx)
+
+    @idx.setter
+    def idx(self, array):
+        self._idx = array
+        self.c_tree.idx = &self._idx[0]
+
+    @property
+    def num_octs(self):
+        return self._num_octs
+
+    @num_octs.setter
+    def num_octs(self, val):
+        self._num_octs = val
+
+    @property
+    def n_ref(self):
+        return self._n_ref
+
+    @n_ref.setter
+    def n_ref(self, val):
+        self._n_ref = val
+
+    @property
+    def num_particles(self):
+        return self._num_particles
+
+    @num_particles.setter
+    def num_particles(self, val):
+        self._num_particles = val
+
+    @property
+    def min_depth(self):
+        return self._min_depth
+
+    @property
+    def leaf_positions(self):
+        cdef int i, j, z = 0
+        positions = []
```

Comment:
Why use a list here? I think it would probably be faster to walk the tree once, count the number of leaf nodes, then allocate a memoryview and fill it while walking the tree again. Filling a python list using `append` has O(N^2) time complexity because it has to copy the whole list every time.

Reply:
yeah, this is just quick sloppy code. Good catch!

Reply:
fixed